p4-16/spec/P4-16-spec.mdk

```diff
@@ -2717,13 +2775,163 @@ X.v1  // reference to v1
 v1    // error - v1 is not in the top-level namespace
 ~ End P4Example
 
-Similar to errors, `enum` expressions only support equality (`==`)
-and inequality (`!=`) comparisons. Expressions whose type is an `enum`
+Similar to errors, `enum` expressions without a specified underlying type only support equality (`==`)
+and inequality (`!=`) comparisons. Expressions whose type is an `enum` without a specified underlying type
 cannot be cast to or from any other type.
 
-Note that if an `enum` value appears in the control-plane API, the
+An `enum` may also specify an underlying type, such as the following:
+
+~ Begin P4Example
+enum bit<8> E {
+  e1 = 0,
+  e2 = 1,
+  e3 = 2
+}
+~ End P4Example
+
+Note that each symbolic value in the enum must map to a unique fixed-width integer
+representation, for example the `enum` in the following example would result in
+error:
+
+~ Begin P4Example
+enum bit<8> Bad {
+  b1 = 0,
+  b2 = 1,  // Error: b2 and b3 both map to 1
+  b3 = 1,  // one must be changed.
+  b4 = 2
+}
+~ End P4Example
+
+An `enum` with an underlying type also supports explicit casts to and from the
+underlying type.  For instance, the following code:
+
+~ Begin P4Example
+bit<8> x;
+E a = E.e2;
+E b;
+
+x = (bit<8>) a; // sets x to 1
+b = (E) x;      // sets e to E.e2
+~ End P4Example
+
+casts `a`, which was initialized to `E.e2` to a `bit<8>`, using the specified
+fixed-width integer representation for `E.e2`, `1`.  The variable `b` is then set to the
+symbolic value `E.e2`, which corresponds to the fixed-width integer value `1`.
+
+Note that while it is always safe to cast from an `enum` to its fixed-width integer type,
+it may not be the case that every fixed-width integer value specified by the underlying type
+has a related symbol in the `enum`.  For instance, the following code:
+
+~ Begin P4Example
+bit<8> x = 5;
+E e = (E) x; // sets e to an unspecified value
+~ End P4Example
+
+sets `e` to an unknown value, since there is no symbol corresponding to the
+fixed-width integer value `5`.  While the value of `e` is unspecified, it must not
+correspond to any symbol defined within enum `E`.  This is because casting from
+a fixed-width integer type to an `enum` type is required to obey the following semantics:
```

Comment:
I know you have the later condition that (bit<W>)(E)x == x for all possible `2^W` values of x of type `bit<W>`.  Assuming we keep that condition (seems reasonable to me), it isn't really true that the assignment to `e` in the example code above assigns `e` an "unknown" value?  I mean, it must be a value not equal to any named member of type `E`, I agree, but it must also be able to be cast back to a `bit<8>` type value of 5, and no other, so we know a lot more about it than that it is simply unknown.

I mention this simply because uninitialized values in P4_16 are, I believe, "more unknown" than the value of `e` in this sample code.